src/rank_llm/rerank/pointwise/monot5.py

```diff
@@ -0,0 +1,133 @@
+import logging
+import math
+from typing import List, Tuple
+
+from transformers import T5ForConditionalGeneration, T5Tokenizer
+from transformers.generation import GenerationConfig
+
+from rank_llm.data import Result
+from rank_llm.rerank.pointwise.pointwise_rankllm import PointwiseRankLLM
+
+logger = logging.getLogger(__name__)
+
+
+class MonoT5(PointwiseRankLLM):
+    def __init__(
+        self,
+        model: str,
+        prompt_mode: str = "monot5",
+        context_size: int = 512,
+        device: str = "cuda",
+        batch_size: int = 32,
+    ):
+        super().__init__(
+            model=model,
+            context_size=context_size,
+            prompt_mode=prompt_mode,
+            device=device,
+            batch_size=batch_size,
+        )
+
+        self._tokenizer = T5Tokenizer.from_pretrained(model)
+        self._llm = T5ForConditionalGeneration.from_pretrained(model).to(self._device)
+        self._context_size = context_size
+
+    def run_llm_batched(
+        self,
+        prompts: List[str],
+    ) -> Tuple[List[str], List[int], List[float]]:
+        gen_cfg = GenerationConfig.from_model_config(self._llm.config)
+        gen_cfg.max_new_tokens = self.num_output_tokens()
+        gen_cfg.min_new_tokens = self.num_output_tokens()
+        gen_cfg.output_scores = True
+        gen_cfg.return_dict_in_generate = True
+        gen_cfg.do_sample = False
+
+        all_outputs = []
+        all_output_token_counts = []
+        all_scores = []
+
+        batch_prompts = prompts
+
+        token_prompts = self._tokenizer(
+            batch_prompts, padding=True, truncation=True, return_tensors="pt"
+        ).to(self._device)
+
+        token_prompts = token_prompts["input_ids"]
+
+        batch_outputs = self._llm.generate(token_prompts, generation_config=gen_cfg)
+
+        batch_output_ids = batch_outputs.sequences
+        batch_logits = batch_outputs.scores
+
+        batch_outputs = [
+            self._tokenizer.decode(
+                single_token_sequence,
+                skip_special_tokens=True,
+                spaces_between_special_tokens=False,
+            )
+            for single_token_sequence in batch_output_ids
+        ]
+
+        for logit_tensor in batch_logits[0]:
+            truth_logit = logit_tensor[1176]
+            false_logit = logit_tensor[6136]
+            score = math.exp(truth_logit) / (
```

Comment:
this is hardcoded and won't scale in the future to other T5 base models, note this as a T5 todo